File: 백준/Gold/2608. 로마 숫자/로마 숫자.py
import sys
input = sys.stdin.readline

rom_to_arab = {'I':1, 'V':5, 'X':10, 'L':50, 'C':100, 'D':500, 'M':1000}
arab_to_rome = {1:'I', 5:'V', 10:'X', 50:'L', 100:'C', 500:'D', 1000:'M'}

def to_arab(s):
    sum = 0

    for i in range(len(s) - 1):
        if rom_to_arab[s[i]] < rom_to_arab[s[i + 1]]:
            sum -= rom_to_arab[s[i]]
        else:
            sum += rom_to_arab[s[i]]
    sum += rom_to_arab[s[-1]]
    return sum

def to_rome(num):
    res = []

    temp = 1000
    while(temp != 0):
        div = num // temp
        c = arab_to_rome[temp]
        if div < 4:
            for _ in range(div):
                res.append(c)
        elif div == 4:
            res.append(c)
            res.append(arab_to_rome[temp * 5])
        elif div < 9:
            res.append(arab_to_rome[temp * 5])
            for _ in range(div - 5):
                res.append(c)
        elif div == 9:
            res.append(c)
            res.append(arab_to_rome[temp * 10])
        num %= temp
        temp //= 10

    return "".join(res)

n1 = input().rstrip()
n2 = input().rstrip()

sum = to_arab(n1) + to_arab(n2)
print(sum)
print(to_rome(sum))

File: 백준/Gold/2608. 로마 숫자/로마 숫자.java
import java.io.BufferedReader;
import java.io.InputStreamReader;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        String n1 = br.readLine();
        String n2 = br.readLine();

        int sum = toArab(n1) + toArab(n2);
        System.out.println(sum);
        System.out.println(toRome(sum));

        br.close();
    }

    public static String toRome(int num) {
        StringBuilder sb = new StringBuilder();

        int temp = 1000;
        while(temp != 0) {
            int div = num / temp;
            char c = toRomeNum(temp);
            if(div < 4) {
                for (int d = 0; d < div; d++) sb.append(c);
            } else if(div == 4) {
                sb.append(c).append(toRomeNum(temp * 5));
            } else if (div < 9) {
                sb.append(toRomeNum(temp * 5));
                for (int d = 0; d < div - 5; d++) sb.append(c);
            } else if(div == 9) {
                sb.append(c).append(toRomeNum(temp * 10));
            }
            num %= temp;
            temp /= 10;
        }

        return sb.toString();
    }

    public static char toRomeNum(int n) {
        switch (n) {
            case 1:
                return 'I';
            case 5:
                return 'V';
            case 10:
                return 'X';
            case 50:
                return 'L';
            case 100:
                return 'C';
            case 500:
                return 'D';
            case 1000:
                return 'M';
            default:
                return 'x';
        }
    }

    public static int toArab(String s) {
        int sum = 0;

        for(int i = 0; i < s.length() - 1; i++) {
            if(toArabNum(s.charAt(i)) < toArabNum(s.charAt(i + 1))) {
                sum -= toArabNum(s.charAt(i));
            } else {
                sum += toArabNum(s.charAt(i));
            }
        }
        sum += toArabNum(s.charAt(s.length() - 1));

        return sum;
    }

    public static int toArabNum(char c) {
        switch (c) {
            case 'I':
                return 1;
            case 'V':
                return 5;
            case 'X':
                return 10;
            case 'L':
                return 50;
            case 'C':
                return 100;
            case 'D':
                return 500;
            case 'M':
                return 1000;
            default:
                return -1;
        }
    }
}
